### PR TITLE
Interface addpoint:change void* to const void*

### PR DESCRIPTION
**Pull request by uestc-lfs**
change from addPoint(void*, args...) to addPoint(const void*, args...).

The changes include the interface in bruteforce.h, and all interfaces
related to addPoint in hnswlib.

I test the code using the 1 million bigann data, the result is ok.

This is from issue #146

### CODE REVIEW

File: hnswlib/hnswlib.h [hunk 1]
```diff
@@ -60,8 +77,8 @@ namespace hnswlib {
     template<typename dist_t>
     class AlgorithmInterface {
     public:
-        virtual void addPoint(void *datapoint, labeltype label)=0;
-        virtual std::priority_queue<std::pair<dist_t, labeltype >> searchKnn(const void *, size_t) const = 0;
+        virtual void addPoint(const void *datapoint, labeltype label)=0;
+        virtual retType<dist_t> searchKnn(const void *, size_t) const = 0;
```
Comment (yurymalkov) [inline hnswlib/hnswlib.h]: Changing the result of the `searchKnn` breaks backward compatibility with all C++ code...
I think adding a separate method with template comparator would be a better solution, isn't it?

**Comment by uestc-lfs** [comment]
Yes,I realized that. I will provide another interface tomorrow.  Besides, will change to const void break the compatibility with python interface?

**Comment by yurymalkov** [comment]
@uestc-lfs I think it is ok to change the python interface as it is an internal part of the library.

File: hnswlib/bruteforce.h [hunk 2]
```diff
@@ -106,6 +109,28 @@ namespace hnswlib {
...
+            if (cur_element_count == 0) return result;
+
+            auto ret = searchKnn(query_data, k);
+
+            while (!ret.empty()) {
+                result.push_back(ret.top());
+                ret.pop();
+            }
+
+            if (result.size() > 1) {
```
Comment (yurymalkov) [inline hnswlib/bruteforce.h]: This additional loop can be avoided if filling the `result` from the back. This doesn't seem to be a big performance issue.

File: hnswlib/hnswalg.h [hunk 3]
```diff
@@ -484,6 +484,8 @@ namespace hnswlib {
 
 
         std::priority_queue<std::pair<dist_t, tableint>> searchKnnInternal(void *query_data, int k) {
+            std::priority_queue<std::pair<dist_t, tableint  >> top_candidates;
+            if (cur_element_count == 0) return top_candidates;
```
Comment (yurymalkov) [inline hnswlib/hnswalg.h]:
```suggestion
            if (cur_element_count == 0) {
                  std::priority_queue<std::pair<dist_t, tableint  >> top_candidates;
                  return top_candidates;
            }
```

Comment (yurymalkov) [reply hnswlib/hnswalg.h]: Not sure if the compiler does not optimize this, but there is no need to create `top_candidates` if it does not branch.

File: hnswlib/bruteforce.h (hunk 2 above)
Comment (uestc-lfs) [reply hnswlib/bruteforce.h]: Not quite understand. Can you show me how to do that?

Comment (yurymalkov) [reply hnswlib/bruteforce.h]: Something like:
```c++           
            if (cur_element_count == 0) {
                std::vector<std::pair<dist_t, labeltype>> result;
                return result;
            }
            
            auto ret = searchKnn(query_data, k);
            
            std::vector<std::pair<dist_t, labeltype>> result(ret.size);
            int i=ret.size-1;
            
            while (!ret.empty()) {
                result[i]=ret.top();
                ret.pop();
                i-=1;
            }
            return result;
```

**Comment by uestc-lfs** [comment]
You assume that ret has a descending order and result has a ascending order. However, I want to return the result according to the order specified by the comparator. And I also assume that the order of ret is unknown.

2019年9月28日 上午12:43，Yury Malkov <notifications@github.com>写道：

@yurymalkov commented on this pull request.

________________________________

In hnswlib/bruteforce.h<https://github.com/nmslib/hnswlib/pull/153#discussion_r329159699>:

> @@ -106,6 +109,28 @@ namespace hnswlib {
             return topResults;
         };

+        template <typename Comp>
+        std::vector<std::pair<dist_t, labeltype>>
+        searchKnn(const void* query_data, size_t k, Comp comp) {
+            std::vector<std::pair<dist_t, labeltype>> result;
+            if (cur_element_count == 0) return result;
+
+            auto ret = searchKnn(query_data, k);
+
+            while (!ret.empty()) {
+                result.push_back(ret.top());
+                ret.pop();
+            }
+
+            if (result.size() > 1) {


Something like:

            if (cur_element_count == 0) {
                std::vector<std::pair<dist_t, labeltype>> result;
                return result;
            }

            auto ret = searchKnn(query_data, k);

            std::vector<std::pair<dist_t, labeltype>> result(ret.size);
            int i=ret.size-1;

            while (!ret.empty()) {
                result[i]=ret.top();
                ret.pop();
                i-=1;
            }
            return result;

**Comment by yurymalkov** [comment]
You are right. 

However, it seems that this logic (check only for the edge elements https://github.com/nmslib/hnswlib/pull/153/files#diff-8ef7e92689e692a09f2a010888a5c7f8R126) works only for two types of comparators (strictly ascending or descending for the first element in tuple). 

This might confuse the user.  I think it might be better to directly use `std::sort`. What do you think?

**Comment by uestc-lfs** [comment]
Yes, std::sort will cover all case.

I'll add it later.

2019年9月28日 上午4:56，Yury Malkov <notifications@github.com>写道：

You are right.

However, it seems that this logic (check only for the edge elements https://github.com/nmslib/hnswlib/pull/153/files#diff-8ef7e92689e692a09f2a010888a5c7f8R126) works only for two types of comparators (strictly ascending or descending for the first element in tuple).

This might confuse the user. I think it might be better to directly use std::sort. What do you think?

**Comment by uestc-lfs** [comment]
@yurymalkov My email shows that there are another two comments:
image
image
But I could not see them on the pull request comments.

Why is putting `top_candidates` inside the `if` block better?

**Comment by yurymalkov** [comment]
@uestc-lfs 
Thanks, will merge after the update.
I've deleted these comments, since I understood they were wrong.

**Comment by yurymalkov** [comment]
@uestc-lfs Thank you!

The tests for the commit failed, but it is cause by a problem in the tests themselves.

**Pull request merged**